sailfishusermanager.cpp

```diff
@@ -0,0 +1,429 @@
+                    qWarning() << "User delete failed:" << lu_strerror(error);
+                    QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, lu_strerror(error)));
+                }
+            } else {
+                qWarning() << "Could not find user";
+                // Error is not populated in this case so returning generic error
+                QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Could not find user"));
+            }
+        } else {
+            qWarning() << "Group delete failed:" << lu_strerror(error);
+            QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, lu_strerror(error)));
+        }
+    } else {
+        qWarning() << "Could not find group";
+        // Error is not populated in this case so returning generic error
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Could not find group"));
+    }
+
+    lu_ent_free(ent);
+    lu_end(context);
+}
+
+void SailfishUserManager::modifyUser(const QString& user, const QString& new_user, const QString& new_name, const QDBusMessage& message)
+{
+    m_timer->start();
+
+    struct lu_error *error = nullptr;
+    struct lu_context *context = lu_start(NULL, lu_user, NULL, NULL, lu_prompt_console, NULL, &error);
+    if (!context) {
+        qWarning() << "Error creating context:" << lu_strerror(error);
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, lu_strerror(error)));
+        return;
+    }
+
+    struct lu_ent *ent = lu_ent_new();
+    if (lu_user_lookup_name(context, user.toLocal8Bit(), ent, &error)) {
+        lu_ent_set_string(ent, LU_USERNAME, new_user.toLocal8Bit());
+        lu_ent_set_string(ent, LU_GECOS, new_name.toLocal8Bit());
+        if (lu_user_modify(context, ent, &error)) {
+            if (lu_group_lookup_name(context, user.toLocal8Bit(), ent, &error)) {
+                lu_ent_set_string(ent, LU_GROUPNAME, new_user.toLocal8Bit());
+                if (lu_group_modify(context, ent, &error)) {
+                    emit userModified(user, new_user, new_name);
+                } else {
+                    qWarning() << "Group modify failed:" << lu_strerror(error);
+                    QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, lu_strerror(error)));
+                }
+            } else {
+                qWarning() << "Could not find group";
+                // Error is not populated in this case so returning generic error
+                QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Could not find group"));
+            }
+        } else {
+            qWarning() << "User modify failed:" << lu_strerror(error);
+            QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, lu_strerror(error)));
+        }
+    } else {
+        qWarning() << "Could not find user";
```

Comment:
matter of taste etc - feel free to ignore, but I find "dealing errors upfront" leads to more readable code than "stacking failure handling in else of else of else ..."

Reply:
matter of taste indeed ) I personally prefer stacking of ifs to returns from the middle - the former also forces separation of meaning pieces of code into relatively small functions, because 4-5 nested ifs is already a bit too much for comfortable reading.

Reply:
True, but I hate copy&pasteing those lu_free* stuff so decided to use this spaghetti approach instead, at least the code is indented :) Adding gotos is also something I refuse to do in C/C++.

Reply:
I think separation of libuser parts and the interface could help with error handling too.